(DOCSP-26981): @realm/reactify - Change an Object Model - React Native (#2399)
mongodb/docs-realm: source/sdk/react-native/realm-database/change-an-object-model.txt

.. tip:: Bypass Migration During Development

   When developing or debugging your application, you may prefer to delete the
   realm instead of migrating it. Use the ``deleteRealmIfMigrationNeeded`` flag
   to delete the database automatically when a schema mismatch requires a

>    migration. 

Reviewer: to call this a "flag" seems incorrect to me. i think calling it "property" would be more precise. also would be good to specify what object this is a member of, i think `Configuration.deleteRealmIfMigrationNeeded`.

also would be good to link to reference docs for this.

schema version of each realm and uses it to map the objects in each
realm to the correct schema.

> Schema versions are ascending integers you can optionally include

Reviewer: ```suggestion
Schema versions are ascending integers you can include
```

Reviewer: https://github.com/mongodb/docs-realm/pull/2399/files#r1050945406

Author: fixed locally

schema version of each realm and uses it to map the objects in each
realm to the correct schema.

Schema versions are ascending integers you can include in the realm
configuration. If a client application does not specify a version number, the
realm defaults to version ``0``.

.. important:: Increment Versions Monotonically
   
   :ref:`Migrations <react-native-migrations>` must update a realm to a higher 

>    schema version. Realm Database will throw an error if a client application 

Reviewer: ```suggestion
   schema version. Realm Database throws an error if a client application 
```

~~~~~~~~~~~~~~

To add a property to a schema, add the new property to the object's class and

> set a ``schemaVersion`` of the ``RealmConfiguration`` object.

Reviewer: it's incorrect to call this RealmConfiguration. the class name is just [Configuration](https://www.mongodb.com/docs/realm-sdks/js/latest/Realm.html#~Configuration)

```suggestion
set a ``schemaVersion`` of the :js-sdk:`Configuration <Realm.html#~Configuration>` object.
```

Author: Fixed locally

Reviewer: this still isn't changed.

Author: Whoops, looks like I may have not pushed up my changes related to that. I've fixed this now, see: https://github.com/mongodb/docs-realm/pull/2399/commits/eee06a2bd8f11868837bd4c9cbe198cc053b2b87

.. example::
   
   A realm using :ref:`schema version
   <react-native-schema-version>` ``1`` has a ``Person`` object type with a
   ``firstName``, and ``lastName`` property. The developer decides to add an
   ``age`` property to the ``Person`` class. 

   To migrate the realm to conform to the updated ``Person`` schema, the
   developer sets the realm's schema version in the ``RealmConfiguration`` to ``2``.
   Finally, the developer passes the configuration object to the

>    ``createRealmContext()`` method.

Reviewer: to refer to user as 'the developer' here reads weird to me, and goes against the mongoDB style guide, which advocates 2nd person and imperative mood - https://www.mongodb.com/docs/meta/style-guide/writing/write-to-the-user/

could you refactor these sentences with this in mind?

Author: I think we're very inconsistent with using "developer" vs "user" throughout our docs, here are just a few examples of where we use developer ([Update a Realm Object Schema - Flutter SDK](https://github.com/mongodb/docs-realm/blob/master/source/sdk/flutter/model-data/update-realm-object-schema.txt#L80), [Change an Object Model - Swift SDK](https://github.com/mongodb/docs-realm/blob/master/source/sdk/swift/model-data/change-an-object-model.txt#L132), [Modify an Object Schema - Node.js SDK](https://github.com/mongodb/docs-realm/blob/master/source/sdk/node/examples/modify-an-object-schema.txt#L37)). 

It's an easy enough change for me to make "developer" -> "user" if you feel particularly strongly about it, but I do think we might prefer to use "developer" for the following reasons:

* Main reason: "User" is ambiguous. Does this refer to the user using the application, or the user of realm (i.e the application developer), or the reader of this docs page (who may be one of many personas) ?
* Supplementary reason: To highlight that this is a specific example of how change the schema. Using the word "developer" indicates that someone else is doing this, not that this is the only way to do it. The 

Also, it seems like the mongoDB style guide is specifically advocating not to use the phrase "the user".

![Screen Shot 2023-01-03 at 11 53 45 AM](https://user-images.githubusercontent.com/52428905/210403423-250f19bd-4152-4d69-a841-4894e8187bcc.png)

Reviewer: ok, but i was commenting that it should be changed to 2nd person ('you') or imperative, not 'user', so i think the initial comment still stands.

Author: Got it, altered to 'you' instead now.

   .. tabs-realm-languages::
     
     .. tab::
         :tabid: typescript
 
         .. literalinclude:: /examples/generated/react-native/ts/change-an-object-model-test.snippet.add-a-property-to-schema.tsx
           :language: typescript
           :emphasize-lines: 4, 12, 20
     
     .. tab::
         :tabid: javascript
 
         .. literalinclude:: /examples/generated/react-native/js/change-an-object-model-test.snippet.add-a-property-to-schema.jsx
           :language: javascript
           :emphasize-lines: 8, 16

.. _react-native-delete-a-property:

Delete a Property
~~~~~~~~~~~~~~~~~

To delete a property from a schema, remove the property from the object's class

> and set a ``schemaVersion`` of the :js-sdk:`RealmConfiguration

Reviewer: again, not the type name

<Realm.html#~Configuration>` object. Deleting a property will not impact
existing objects.

.. example::

   A realm using :ref:`schema version <react-native-schema-version>` ``1`` has a
   ``Person`` object type with a ``lastName`` property. The developer decides to
   remove the property from the schema. 

   To migrate the realm to conform to the updated ``Person`` schema, the
   developer sets the realm's schema version to 2 in the ``RealmConfiguration``
   object. Finally, the developer passes the configuration object to the

>    ``createRealmContext()`` method.

Reviewer: same as above about 'the developer'

Author: Addressed here: https://github.com/mongodb/docs-realm/pull/2399#discussion_r1060772892

~~~~~~~~~~~~~~

To add a property to a schema, add the new property to the object's class and
set a ``schemaVersion`` of the ``Configuration`` object.

.. example::
   
   A realm using :ref:`schema version
   <react-native-schema-version>` ``1`` has a ``Person`` object type with a
   ``firstName`` and ``lastName`` property. You decide to add an
   ``age`` property to the ``Person`` class. 

   To migrate the realm to conform to the updated ``Person`` schema, the

>    developer sets the realm's schema version in the ``Configuration`` to ``2``.

Reviewer: change to 2nd person

>    Finally, you pass the configuration object to the

Reviewer: ```suggestion
   Finally, pass the configuration object to the
```

   ``createRealmContext()`` method.

   .. tabs-realm-languages::
     
     .. tab::
         :tabid: typescript
 
         .. literalinclude:: /examples/generated/react-native/ts/change-an-object-model-test.snippet.add-a-property-to-schema.tsx
           :language: typescript
           :emphasize-lines: 4, 12, 20
     
     .. tab::
         :tabid: javascript
 
         .. literalinclude:: /examples/generated/react-native/js/change-an-object-model-test.snippet.add-a-property-to-schema.jsx
           :language: javascript
           :emphasize-lines: 8, 16

.. _react-native-delete-a-property:

Delete a Property
~~~~~~~~~~~~~~~~~

To delete a property from a schema, remove the property from the object's class
and set a ``schemaVersion`` of the :js-sdk:`Configuration
<Realm.html#~Configuration>` object. Deleting a property will not impact
existing objects.

.. example::

   A realm using :ref:`schema version <react-native-schema-version>` ``1`` has a
   ``Person`` object type with a ``lastName`` property. You decide to
   remove the property from the schema. 

>    To migrate the realm to conform to the updated ``Person`` schema, the
>    developer sets the realm's schema version to 2 in the ``Configuration``

Reviewer: ```suggestion
   To migrate the realm to conform to the updated ``Person`` schema,
   set the realm's schema version to 2 in the ``Configuration``
```

>    object. Finally, you pass the configuration object to the

Reviewer: ```suggestion
   object. Finally, pass the configuration object to the
```

~~~~~~~~~~~~~~~~~

To rename an object property, change the property name in the object schema and
then create a realm configuration with an incremented :ref:`schema version
<react-native-schema-version>` and a :ref:`migration <react-native-migrations>`
function that updates existing objects to use the new property name.

Migrations do not allow you to directly rename a property. Instead, you can
create a new property with the updated name, copy the value from the old
property, and then delete the old property.

.. example::
   
   A realm using :ref:`schema version <react-native-schema-version>` ``1`` has
   a ``Person`` object type. The original schema had a ``firstName`` and ``lastName``
   field. The developer later decides that the ``Person`` class should use a combined
   ``fullName`` field and removes the separate ``firstName`` and ``lastName`` fields.
   
   To migrate the realm to conform to the updated ``Person`` schema, the

>    developer creates a :js-sdk:`RealmConfiguration <Realm.html#~Configuration>` object

Reviewer: again about 'the developer'

Author: Addressed here: https://github.com/mongodb/docs-realm/pull/2399#discussion_r1060772892

~~~~~~~~~~~~~~~~~

To rename an object property, change the property name in the object schema and
then create a realm configuration with an incremented :ref:`schema version
<react-native-schema-version>` and a :ref:`migration <react-native-migrations>`
function that updates existing objects to use the new property name.

Migrations do not allow you to directly rename a property. Instead, you can
create a new property with the updated name, copy the value from the old
property, and then delete the old property.

.. example::
   
   A realm using :ref:`schema version <react-native-schema-version>` ``1`` has
   a ``Person`` object type. The original schema had a ``firstName`` and ``lastName``
   field. You later decide that the ``Person`` class should use a combined
   ``fullName`` field and removes the separate ``firstName`` and ``lastName`` fields.
   

>    To migrate the realm to conform to the updated ``Person`` schema, you create a :js-sdk:`Configuration <Realm.html#~Configuration>` object

Reviewer: ```suggestion
   To migrate the realm to conform to the updated ``Person`` schema, create a :js-sdk:`Configuration <Realm.html#~Configuration>` object
```

   and set the realm's schema version to ``2``, and define a migration function that
   sets the value of ``fullName`` based on the existing ``firstName`` and ``lastName``

>    properties. Finally, you pass the configuration object to the

Reviewer: ```suggestion
   properties. Finally, pass the configuration object to the
```

~~~~~~~~~~~~~~~~~~~~~~
To modify a property's type, set the property type of the field that you wish to
modify to the new data type. Then, set a ``schemaVersion``, and a ``migration``

> callback function of the :js-sdk:`RealmConfiguration

Reviewer: same as above about `RealmConfiguration`

<Realm.html#~Configuration>` object. 

.. note::

   :ref:`Synchronized realms <sync>` only support non-breaking changes
   to ensure that older clients can sync with newer clients. This means that
   synchronized realms do not support modifying the type of a property of a
   schema. 

.. example::
   
   A realm using :ref:`schema version <react-native-schema-version>` ``1`` has a
   ``Task`` object type. The original schema had an ``_id`` with a property type of
   ``int``. You later decide that the ``Task`` class's ``_id`` field should
   be of type ``ObjectId``, and updates the schema.

>    To migrate the realm to conform to the updated ``Task`` schema, you create a

Reviewer: ```suggestion
   To migrate the realm to conform to the updated ``Task`` schema, create a
```

>    creates a :js-sdk:`RealmConfiguration <Realm.html#~Configuration>` object and

Reviewer: same as above about `RealmConfiguration`

   define a migration function to convert the integer type to an ``Object ID``

>    type. Finally, you pass the configuration object to the

Reviewer: ```suggestion
   type. Finally, pass the configuration object to the
```

~~~~~~~~~~~~~~~~~~~~~~
To modify a property's type, set the property type of the field that you wish to
modify to the new data type. Then, set a ``schemaVersion``, and a ``migration``
callback function of the :js-sdk:`Configuration
<Realm.html#~Configuration>` object. 

.. note::

   :ref:`Synchronized realms <sync>` only support non-breaking changes
   to ensure that older clients can sync with newer clients. This means that
   synchronized realms do not support modifying the type of a property of a
   schema. 

.. example::
   
   A realm using :ref:`schema version <react-native-schema-version>` ``1`` has a
   ``Task`` object type. The original schema had an ``_id`` with a property type of
   ``int``. The developer later decides that the ``Task`` class's ``_id`` field should
   be of type ``ObjectId``, and updates the schema.

   To migrate the realm to conform to the updated ``Task`` schema, the developer

>    creates a :js-sdk:`Configuration <Realm.html#~Configuration>` object and

Reviewer: [suggestion] don't make this a link b/c you link to this just a few lines above

https://github.com/mongodb/docs-realm/pull/2399/files#diff-e4287620889b14ddb45f89cb33481d6ca1eff15446ff3d03b465e670f8707025R228-R229
